### DISCUSSION TITLE
I made a program converts to Reduced Row Echelon Form (RREF) for Linear Algebra. But something wrong(Python)

### DISCUSSION THREAD

**Discussion by MPJUNS** [community/community#114448] [Programming Help]
It works when problem with ordered pairs but when problem comes with matrix, the result is different. What part of function is wrong? I asked at Chatgpt, but I couldn't get clear answer. Can you guys help me?

Please understand about my english writing. I'm from korea, so I can only use basic english. Thanks you
### Body

`# Program V3.0 that converts to Reduced Row Echelon Form (RREF) for Linear Algebra
# Function call for fraction representation
from fractions import Fraction 

# Function for matrix output
def testfunc(ans):
    for para in ans:
        print([str(Fraction(factor).limit_denominator()) for factor in para]) 
    print()

# Part to store (ordered pair - 1)st order equation coefficients as a list
def MakeVarfunc(poly):
    varlist = []
    polyans2di = [] 
    for i in range(poly):
        var = list(map(float,input("Enter the values of x{}, y{}:".format(i+1,i+1)).split())) 
        varlist.append(var)
    for i in range(len(varlist)):
        polyans1di = [1] 
        for k in range(1,poly):
            polyans1di.append((varlist[i][0])**k)
        for k in range(1):
            polyans1di.append(varlist[i][1])
        polyans2di.append(polyans1di)
    return polyans2di

# Function for matrix representation
def MakeMatrixfunc(m,n):
    dimen1 = [] 
    for row in range(m):
        dimen2 = [] 
        dimen2 = list(map(float,input("Enter space-separated numbers for row {}:".format(row+1)).split()))
        dimen1.append(dimen2)
    print("The entered matrix is:")
    testfunc(dimen1)
    print("Here it is...!")
    return dimen1

# Part to convert a list (ordered pairs) to RREF
def RREFfunc_num(poly):
    ans = MakeVarfunc(poly)
    print("Results after arranging the coefficients of equations into a matrix:")
    testfunc(ans)
    print("Conversion starts....")
    print("Arranging for a0 into RREF....")
    lange = len(ans)
    for a0 in range(1,lange):
        count = 1
        print("R{} - R0 -> R{}".format(a0,a0))
        ans[a0] = [b - a for b,a in zip(ans[a0],ans[0])]
        testfunc(ans)
    for standard in range(1,poly):
        print("Arranging for a{} into RREF....".format(standard)) 
        divide = ans[standard][standard]
        if divide != 0:
            count += 1
            print("(1/{})R{} -> R{}".format(divide, standard, standard))
            testfunc(ans)
            ans[standard] = [b / divide for b in ans[standard]]
            for sub in range(poly):
                if sub != standard:
                    nomal = ans[sub][standard]
                    print("R{} - ({})R{} -> R{}".format(sub, nomal,standard,sub))
                    ans[sub] = [b - (a*nomal) for b, a in zip(ans[sub],ans[standard])]
                    testfunc(ans)
        else:
            print("Skipping RREF for coefficient a{} because it's 0 in {}th row.".format(standard,standard))
            continue
    return ans 

# Part to convert a list (matrix) to RREF
def RREFfunc_Matrix(Matrix):
    poly = len(Matrix)
    for standard in range(poly):
        print("Arranging for a{} into RREF....".format(standard)) 
        if Matrix[standard][standard] == 0:
            find = 0
            while find < standard and Matrix[standard][find] != 0: 
                find += 1
            if find == standard :
                continue
            Matrix [standard], Matrix[find] = Matrix[find] , Matrix[standard]
        divide = Matrix[standard][standard]
        if divide != 0:
            print("(1/{})R{} -> R{}".format(divide, standard, standard))
            testfunc(Matrix)
            Matrix[standard] = [b / divide for b in Matrix[standard]]
            for sub in range(poly):
                if sub != standard and sub != poly - 1:
                    nomal = Matrix[sub][standard]
                    print("R{} - ({})R{} -> R{}".format(sub, nomal,standard,sub))
                    Matrix[sub] = [b - (a*nomal) for b, a in zip(Matrix[sub],Matrix[standard])]
                    testfunc(Matrix)
        else:
            print("Skipping RREF for coefficient a{} because it's 0 in {}th row.".format(standard,standard))
            continue
    return Matrix     

# Function to print coefficient values
def printvaluefunc(poly, ans):
    for standard in range(poly):
       print("a{} = {}".format(standard, str(Fraction(ans[standard][-1]).limit_denominator())))

# Main function part
def Main():
    for i in range(60):
        print("*",end = "")
    print("\n")
    print("Gauss Jordan Elimination Program V3.0 by Kim Jun-seong")
    print("\n")
    for i in range(60):
        print("*",end = "")
    print("\n")
    print("Choose the data format.")
    choice = 0
    while choice < 1 or choice > 2:
        choice = int(input("Ordered pair format: 1 Matrix format: 2"))
        if choice == 1:
            poly = int(input("How many ordered pairs are there?")) 
            print("When you form equations..... it's a {}rd degree equation with coefficients a0 ~ a{}..!".format(poly-1,poly-1))
            ans = RREFfunc_num(poly) 
            printvaluefunc(poly,ans)
            print("\n")
            print("Program ends....")
            break
        elif choice == 2:
            m, n = map(int,input("Enter the size of the matrix [in the form m,n]:").split(","))
            if m == n or m > n:
                print("Sorry, it's a format that cannot be solved because there are one or more equations more than the variables.")
                break
            else:
                Matrix = MakeMatrixfunc(m,n)
                ans = RREFfunc_Matrix(Matrix)
                printvaluefunc(m,ans)
                print("\n")
                print("Program ends....")
                break
        else:
            print("Incorrect input.")
            continue

Main() # Program execution`

### Guidelines

- [X] I have read and understood this category's [guidelines](https://github.com/orgs/community/discussions/51384) before making this post.

**Comment by MarouaneBenbetka** [comment]
Try this code :

```
from fractions import Fraction

# Function to print matrix in fraction form


def print_matrix(matrix):
    for row in matrix:
        print([str(Fraction(elem).limit_denominator()) for elem in row])
    print()

# Function to input matrix from user


def input_matrix(prompt):
    print(prompt)
    m, n = map(int, input(
        "Enter the number of rows and columns [m,n]: ").split(','))
    matrix = [list(map(float, input(
        f"Enter space-separated numbers for row {row + 1}: ").split())) for row in range(m)]
    return matrix, m, n

# Function to swap rows


def swap_rows(matrix, i, j):
    matrix[i], matrix[j] = matrix[j], matrix[i]

# Function to reduce matrix to RREF


def rref(matrix):
    row_count = len(matrix)
    col_count = len(matrix[0])

    lead = 0
    for r in range(row_count):
        if lead >= col_count:
            return
        i = r
        while matrix[i][lead] == 0:
            i += 1
            if i == row_count:
                i = r
                lead += 1
                if col_count == lead:
                    return
        swap_rows(matrix, i, r)
        lv = matrix[r][lead]
        matrix[r] = [mrx / float(lv) for mrx in matrix[r]]

        for i in range(row_count):
            if i != r:
                lv = matrix[i][lead]
                matrix[i] = [iv - lv * rv for rv,
                             iv in zip(matrix[r], matrix[i])]
        lead += 1

# Function to print the solutions


def print_solutions(matrix):
    for i, row in enumerate(matrix):
        print(f"a{i} = {Fraction(row[-1]).limit_denominator()}")

# Main function to run the program


def main():
    print("Gauss Jordan Elimination Program V3.0 by Kim Jun-seong\n")
    choice = int(
        input("Choose the data format - Ordered pair format: 1, Matrix format: 2: "))

    if choice == 1:
        # For ordered pairs, additional implementation needed based on your requirements
        print("Ordered pair functionality not implemented in this correction.")
    elif choice == 2:
        matrix, m, n = input_matrix("Enter the matrix for RREF conversion:")

        print("Original Matrix:")
        print_matrix(matrix)

        rref(matrix)
        print("Matrix after RREF conversion:")
        print_matrix(matrix)

        print_solutions(matrix)
    else:
        print("Incorrect input.")


if __name__ == "__main__":
    main()

```
